### PR TITLE
Read cp filters from configuration file

### PR DESCRIPTION
**Pull request by asvetlov**

### CODE REVIEW

**Comment by asvetlov** [comment]
Need to document new config parameter before the merge

**Comment by asvetlov** [comment]
Documented

File: neuromation/cli/job.py [hunk 1]
```diff
@@ -479,8 +479,10 @@ def job() -> None:
     "--format",
     type=JOB_COLUMNS,
     help=(
-        'Output table format, use "neuro help ps-format" '
+        'Output table format, see "neuro help ps-format" '
         "for more info about the format specification."
```
Comment (serhiy-storchaka) [inline neuromation/cli/job.py]:
```suggestion
        "for more info about the format specification. "
```

File: neuromation/api/file_filter.py [hunk 2]
```diff
@@ -18,7 +18,7 @@ def exclude(self, pattern: str) -> None:
     def include(self, pattern: str) -> None:
         self.append(False, pattern)
 
-    def match(self, path: str) -> bool:
+    async def match(self, path: str) -> bool:
```
Comment (serhiy-storchaka) [inline neuromation/api/file_filter.py]: Why make it asynchronous?

File: neuromation/api/config.py [hunk 3]
```diff
@@ -365,6 +369,52 @@ def _check_sections(
...
+    section: str,
+    params: Dict[str, Any],
+    filename: Union[str, "os.PathLike[str]"],
+) -> None:
+    sec = config.get(section)
+    if sec is None:
+        return
+    diff = sec.keys() - params.keys()
+    if diff:
+        diff_str = ", ".join(f"{section}.{name}" for name in diff)
```
Comment (serhiy-storchaka) [inline neuromation/api/config.py]:
```suggestion
        diff_str = ", ".join(f"{section}.{name}" for name in sorted(diff))
```
The order in a set is not defined.

File: neuromation/api/config.py [hunk 4]
```diff
@@ -365,6 +369,52 @@ def _check_sections(
...
+                f"{validator[0].__name__} is expected"
+            )
+        item_type = validator[1]
+        for num, i in enumerate(val):
+            _check_item(i, item_type, f"{full_name}[{num}]", filename)
+    else:
+        assert isinstance(validator, type) and issubclass(
+            validator, (bool, numbers.Real, numbers.Integral, str)
+        )
+        # validator for integer types should be numbers.Real or numbers.Integral,
```
Comment (serhiy-storchaka) [inline neuromation/api/config.py]: Why?

File: neuromation/api/config.py [hunk 5]
```diff
@@ -365,6 +369,52 @@ def _check_sections(
             )
 
 
+def _check_item(
+    val: Any, validator: Any, full_name: str, filename: Union[str, "os.PathLike[str]"],
+) -> None:
+    if isinstance(validator, tuple):
+        # (container_type, item_type) pair
```
Comment (serhiy-storchaka) [inline neuromation/api/config.py]: Maybe add `container_type, item_type = validator`?

File: tests/cli/test_storage.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+) -> None:
+
+    async with make_client("https://example.com") as client:
+        monkeypatch.chdir(tmp_path)
+        local_conf = tmp_path / ".neuro.toml"
+        # empty config
+        local_conf.write_text(
+            toml.dumps({"storage": {"cp-exclude": ["*.jpg", "!main.jpg"]}})
+        )
+        assert calc_filters(client, None) == [(True, ".jpg"), (False, "main.jpg")]
```
Comment (serhiy-storchaka) [inline tests/cli/test_storage.py]:
```suggestion
        assert calc_filters(client, None) == [(True, "*.jpg"), (False, "main.jpg")]
```

File: neuromation/cli/storage.py [hunk 7]
```diff
@@ -856,3 +866,20 @@ def filter_option(*args: str, flag_value: bool, help: str) -> Callable[[Any], An
...
+def calc_filters(
+    client: Client, filters: Optional[Tuple[Tuple[bool, str], ...]]
+) -> Tuple[Tuple[bool, str], ...]:
+    if filters is not None:
+        return filters
+    ret = []
+    config = client.config.get_user_config()
+    section = config.get("storage")
+    if section is not None:
+        for flt in section.get("cp-exclude"):
```
[the remark this answers is not in the document]
Comment (asvetlov) [reply neuromation/cli/storage.py]: You are right, fixing

File: neuromation/api/config.py (hunk 4 above)
Comment (asvetlov) [reply neuromation/api/config.py]: `isinstance(4, float) == False` but `isinstance(4, numbers.Real) == True`

**Comment by asvetlov** [comment]
@serhiy-storchaka thanks for the review!

**Pull request merged**